notes.md

```diff
@@ -0,0 +1,179 @@
+- We want to limit the number of TCP connections? To protect the server?
+    - Backpressure accepting connections (and reads) if memory is getting high?
+      E.g. Netty thread pool is queuing too many tasks in compute thread pool. Queue will hit OOM.
+
+## Implementation thoughts for above use cases
+Based on what currently exists and past discussions with additional modifications.
+
+## Package Structure
+- fs2-netty-core
+  - Netty
+    - wrappers
+    - ops
+- fs2-netty-server
+  - tcp
+  - udp
+- fs2-netty-client
+  - tcp
+  - udp
+- fs2-netty-all
+- fs2-netty-simple-http-server
+  - websockers
+
+### API
+We'll look at how intermediate library/module authors will use FS2-Netty to build protocols for above use-cases. 
+
+#### TCP Server
+The idea is that FS2-Netty users provide a custom Netty pipeline and probably some configs to `FS2TcpServer`, which
+builds the server resource that exposes a Stream of `Socket`'s. Then users can map over that to produce their desired
+api. `Socket` should be the bridge out of Netty, i.e. no more Netty further exposed (maybe this isn't practical,
+there's also more things Netty does besides sockets, so api will be refined). 
+```scala
+// similar to SocketHandler, a bridge out of Netty that exposes streams for I/O types and other TCP connection api methods 
+class NettyPipeline[I, O](handlers: List[NettyHandler])
+
+object FS2TcpServer {
+  def make[F[_], I, O](pipeline: NettyPipeline[I, O]): Resource[F, Fs2Stream[F, ServerSocket[F, I, O]]] = ???
+}
+
+object MySimpleHttpServer {
+
+  private trait HttpSocket[F[_]] {
+    def read(f: HttpRequest => F[HttpResponse]): F[Unit]
+  }
+
+  private[this] val httpPipeline = new NettyPipeline[HttpRequest, HttpResponse](handlers = Nil) // This would contain actual handlers
+
+  def make[F[_]](): Resource[F, Fs2Stream[F, HttpSocket]] =
+    FS2TcpServer.make(httpPipeline)
+      .map(stream => stream.map(tcpServerSocket => new HttpSocket[F] {
+        override def read(f: HttpRequest => F[HttpResponse]): F[Unit] = ??? // interface with tcpServerSocket
+      }))
+}
+```
+
+# Random extras
+```scala
+abstract class WebSocketFrame(underlying: NettyWebSocketFrame)
+trait DataFrame extends WebSocketFrame
+trait ControlFrame extends WebSocketFrame
+
+final case class BinaryFrame() extends DataFrame
+final case class TextFrame() extends DataFrame
+// Pipeline Framework may also handle this, but we should expose for advanced usage 
+final case class ContinuationFrame() extends DataFrame
+
+final case class PingFrame() extends ControlFrame
+final case class PongFrame() extends ControlFrame
+final case class CloseFrame() extends ControlFrame
+
+trait CloseReason
+final case class ConnectionError(cause: Throwable) extends CloseReason
+case object TcpConnectionDropped extends CloseReason
+final case class FrameworkInitiated(closeFrame: CloseFrame) extends CloseReason
+final case class UserInitiated(closeFrame: CloseFrame) extends CloseReason
+
+// Or Observer
```

Comment:
So this is actually where things will get SUPER awesome, because for websockets, I think we'll just expose a `Socket[F, WebSocketFrame, WebSocketFrame]`. So in other words, a socket where you can send and receive websocket frames, just as how without any pipeline handlers we get a socket where you can send and receive bytes, and with the http handler we get a socket where you can send and receive `HttpRequest`/`HttpResponse`, respectively.

Reply:
Yup